From 821e56d51e23543ba505dea2916bb1802531a4fd Mon Sep 17 00:00:00 2001
From: Tony Camuso <tcamuso@redhat.com>
Date: Fri, 8 Aug 2014 18:14:41 -0400
Subject: [char] ipmi: Cleanup error return

Message-id: <1407521687-4858-8-git-send-email-tcamuso@redhat.com>
Patchwork-id: 87548
O-Subject: [RHEL7.1 PATCH 07/13] ipmi: Cleanup error return
Bugzilla: 1111214
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1111214

Cherry-picked from the following upstream commit with PARISC hunk
removed.

commit d02b3709ff8efebfca0612d0ac2a6e31a91c13f4
Author: Corey Minyard <cminyard@mvista.com>
Date:   Fri Jan 24 14:00:53 2014 -0600

    ipmi: Cleanup error return

    Return proper errors for a lot of IPMI failure cases.  Also call
    pci_disable_device when IPMI PCI devices are removed.

    Signed-off-by: Corey Minyard <cminyard@mvista.com>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Signed-off-by: Tony Camuso <tcamuso@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/char/ipmi/ipmi_si_intf.c b/drivers/char/ipmi/ipmi_si_intf.c
index 2312a82..efda0da 100644
--- a/drivers/char/ipmi/ipmi_si_intf.c
+++ b/drivers/char/ipmi/ipmi_si_intf.c
@@ -1841,11 +1841,15 @@ static int hotmod_handler(const char *val, struct kernel_param *kp)
     info->irq_setup = std_irq_setup;
    info->slave_addr = ipmb;
 
-   if (!add_smi(info)) {
-    if (try_smi_init(info))
-     cleanup_one_si(info);
-   } else {
+   rv = add_smi(info);
+   if (rv) {
     kfree(info);
+    goto out;
+   }
+   rv = try_smi_init(info);
+   if (rv) {
+    cleanup_one_si(info);
+    goto out;
    }
   } else {
    /* remove */
@@ -2059,6 +2063,7 @@ struct SPMITable {
 static int try_init_spmi(struct SPMITable *spmi)
 {
  struct smi_info  *info;
+ int rv;
 
  if (spmi->IPMIlegacy != 1) {
   printk(KERN_INFO PFX "Bad SPMI legacy %d\n", spmi->IPMIlegacy);
@@ -2133,10 +2138,11 @@ static int try_init_spmi(struct SPMITable *spmi)
    info->io.addr_data, info->io.regsize, info->io.regspacing,
    info->irq);
 
- if (add_smi(info))
+ rv = add_smi(info);
+ if (rv)
   kfree(info);
 
- return 0;
+ return rv;
 }
 
 static void spmi_find_bmc(void)
@@ -2170,6 +2176,7 @@ static int ipmi_pnp_probe(struct pnp_dev *dev,
  acpi_handle handle;
  acpi_status status;
  unsigned long long tmp;
+ int rv;
 
  acpi_dev = pnp_acpi_device(dev);
  if (!acpi_dev)
@@ -2251,10 +2258,11 @@ static int ipmi_pnp_probe(struct pnp_dev *dev,
    res, info->io.regsize, info->io.regspacing,
    info->irq);
 
- if (add_smi(info))
-  goto err_free;
+ rv = add_smi(info);
+ if (rv)
+  kfree(info);
 
- return 0;
+ return rv;
 
 err_free:
  kfree(info);
@@ -2558,16 +2566,20 @@ static int ipmi_pci_probe(struct pci_dev *pdev,
   &pdev->resource[0], info->io.regsize, info->io.regspacing,
   info->irq);
 
- if (add_smi(info))
+ rv = add_smi(info);
+ if (rv) {
   kfree(info);
+  pci_disable_device(pdev);
+ }
 
- return 0;
+ return rv;
 }
 
 static void ipmi_pci_remove(struct pci_dev *pdev)
 {
  struct smi_info *info = pci_get_drvdata(pdev);
  cleanup_one_si(info);
+ pci_disable_device(pdev);
 }
 
 static struct pci_device_id ipmi_pci_devices[] = {
@@ -2662,9 +2674,10 @@ static int ipmi_probe(struct platform_device *dev)
 
  dev_set_drvdata(&dev->dev, info);
 
- if (add_smi(info)) {
+ ret = add_smi(info);
+ if (ret) {
   kfree(info);
-  return -EBUSY;
+  return ret;
  }
 #endif
  return 0;
-- 
1.7.1